src/messaging/ExchangeContext.cpp

```diff
@@ -573,6 +573,17 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
         return CHIP_NO_ERROR;
     }
 
+    if (IsMessageNotAcked())
+    {
+        // The only way we can get here is a spec violation on the other side:
+        // we sent a message that needs an ack, and the other side responded
+        // with a message that does not contain an ack for the message we sent.
+        // Just drop this message; if we delivered it to our delegate it might
+        // try to send another message-needing-an-ack in response, which would
+        // violate our internal invariants.
+        return CHIP_ERROR_INCORRECT_STATE;
```

Comment:
How common is this? Should we log for easier debugging or does that add a potential DOS point?

Reply:
I found the issue while developing a from scratch prototype, and this is only triggered by an incomplete implementation that is acting in an invalid, non-spec compliant way (so sufficiently rare).

Reply:
Not common.  I will add a log once the unit test bits are sorted.

Reply:
Added a log.